File: packages/gg-sandbox-python/src/gg/delegation.py
"""Delegate work to child agents, and hand this session's own turn to another agent.

`wait_for_subagents` blocks while children run, and the run's budget keeps ticking while it does.

A child is briefed with exactly one of a self-contained `prompt` or a board `issue_id`; neither and
both are refused.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from wit_world.imports import delegation as wire

from ._registry import alias, missing, operation
from .core import ApiError, ApiErrorCode, _call, _strings

__all__ = [
    "AgentEnding",
    "SubagentHandle",
    "SubagentResult",
    "exec",
    "fork",
    "send_message",
    "spawn_subagent",
    "transition_state",
    "wait_for_subagents",
]


@dataclass(frozen=True)
class SubagentHandle:
    """A child agent that was spawned and is now running in parallel."""

    id: str
    """The child's id — what `wait_for_subagents` and `send_message` take."""

    slot: str
    """The agent profile it runs as."""

    model_id: str
    """The model actually bound to that agent."""

    @alias("delegation.send_message")
    def send(self, message: str) -> None:
        """Deliver a message to this child's inbox, which it reads at its next turn.

        Args:
            message: What to put in its inbox.

        Raises:
            ApiError: `conflict` when this child has already returned.
        """
        send_message(self.id, message)


class AgentEnding(Enum):
    """How a child agent's loop ended."""

    COMPLETED = "completed"
    """It finished normally, ending its own session, and its summary is what it returned."""

    EXHAUSTED = "exhausted"
    """It reached the per-run turn ceiling."""

    TIMED_OUT = "timed_out"
    """It passed its wall-clock deadline."""

    MODEL_ERROR = "model_error"
    """A model turn failed."""

    AUTH_ERROR = "auth_error"
    """The run's credential was refused."""

    LIMIT_EXCEEDED = "limit_exceeded"
    """An execution ceiling stopped it — consecutive errors, error rate, or cost."""


@dataclass(frozen=True)
class SubagentResult:
    """One child agent's collected result."""

    id: str
    """The child's id."""

    status: AgentEnding | None
    """How it finished; `None` when it produced no return value at all."""

    summary: str
    """Its final message."""


def _brief(fn: str, prompt: str | None, issue_id: str | None) -> wire.SubagentBrief:
    """Lower a brief onto the membrane's variant.

    The check is explicit rather than delegated to the bindings because "neither" is the mistake that
    actually happens — the native JSON schema declares both fields optional and enforces the choice
    only at dispatch — and the message that comes back has to name both options.
    """
    if (prompt is None) == (issue_id is None):
        raise ApiError(
            fn,
            ApiErrorCode.INVALID_ARGUMENT,
            "expected exactly one of `prompt` or `issue_id`",
        )
    return (
        wire.SubagentBrief_Prompt(prompt)
        if prompt is not None
        else wire.SubagentBrief_Issue(issue_id)
    )


def _handle(handle: wire.SubagentHandle) -> SubagentHandle:
    """The membrane's handle record, as the model-facing one."""
    return SubagentHandle(id=handle.id, slot=handle.slot, model_id=handle.model_id)


@operation("delegation.spawn_subagent")
def spawn_subagent(
    agent: str, *, prompt: str | None = None, issue_id: str | None = None
) -> SubagentHandle:
    """Delegate scoped work to a child agent, which runs in parallel while the program continues.

    `agent` names one of the agent profiles this session may spawn — the system prompt lists them,
    and the profile selects the child's model, tools and instructions. The brief is exactly one of
    `prompt` and `issue_id`. The child shares the workspace.

    Args:
        agent: The agent profile to run the child as, from the ones this session may spawn. It
            selects the child's model, tools and instructions.
        prompt: Self-contained instructions for the child. Give this or `issue_id`, never both and
            never neither.
        issue_id: The board issue to brief the child from. Give this or `prompt`, never both and
            never neither.

    Returns:
        The child's handle: the id `wait_for_subagents` and `send_message` take, the agent profile
            it runs as, and the model bound to that profile.

    Raises:
        ApiError: `limit-exceeded` at the delegation depth cap, and `invalid-argument` when `agent`
            is not one this session may spawn.
    """
    return _handle(
        _call(
            wire.spawn_subagent,
            wire.SpawnRequest(agent=agent, task=_brief("spawn_subagent", prompt, issue_id)),
        )
    )


@operation("delegation.wait_for_subagents")
def wait_for_subagents(ids: list[str] | None = None) -> list[SubagentResult]:
    """Block until the named children have finished and collect their results in dispatch order.

    The default waits for every outstanding child. The run's wall-clock budget keeps running
    throughout.

    Args:
        ids: The children to wait for, as `spawn_subagent` returned them. The default waits for every
            one still outstanding.

    Returns:
        One result per child, in dispatch order, each carrying its final message and how it ended. A
            child that produced no return value at all has no `status`.

    Raises:
        ApiError: `not-found` for an id this session did not spawn.
    """
    waited = None if ids is None else _strings("wait_for_subagents", "ids", ids)
    return [
        SubagentResult(
            id=result.id,
            status=None if result.status is None else AgentEnding[result.status.name],
            summary=result.summary,
        )
        for result in _call(wire.wait_for_subagents, waited)
    ]


@operation("delegation.send_message")
def send_message(agent_id: str, message: str) -> None:
    """Deliver a message to a running child agent's inbox, which it reads at its next turn.

    Args:
        agent_id: The child to deliver to, as `spawn_subagent` returned it.
        message: What to put in its inbox. It reads it at its next turn.

    Raises:
        ApiError: `not-found` for an unknown agent id, and `conflict` when that child has already
            returned.
    """
    _call(wire.send_message, agent_id, message)


@operation("delegation.transition_state")
def transition_state(state: str, note: str | None = None) -> None:
    """Move the process this session is running inside on to another of its states.

    The state is named the way an agent to spawn is named. It is bound only when a state machine is
    driving the session and the current state has somewhere to go. It is registered rather than
    performed: the call validates the target, returns, and the program runs on to its end. The first
    declaration in a turn is the one that stands.

    Args:
        state: The state to move on to, named the way an agent to spawn is named.
        note: The opening message the next state's agent sees. The default tells it nothing.

    Raises:
        ApiError: `invalid-argument` for a state this session may not move to, `refused` for a
            second declaration in one turn, and `unavailable` when this session is not running inside
            a state machine at all.
    """
    _call(wire.transition_state, state, note)


@operation("delegation.exec")
def exec(agent: str, prompt: str | None = None) -> None:
    """Continue this session as a different agent, from the next turn.

    The named agent takes over with its own model, tools and instructions, keeping every capability
    the two of them share and the whole conversation. Registered rather than performed: the call
    validates the target, returns, and the program runs on to its end. A session makes one succession
    per turn, and none at all while a state machine is driving the session.

    Args:
        agent: The agent to become, from the ones this session may become.
        prompt: Its opening message. It already has the whole conversation, so this is the
            instruction rather than a briefing. The default tells it nothing.

    Raises:
        ApiError: `invalid-argument` for an agent this session may not become, `refused` for a
            second succession in one turn, and `unavailable` when this session is running inside a
            state machine, which is left by moving it on to another of its states instead.
    """
    _call(wire.exec, agent, prompt)


@operation("delegation.fork")
def fork(prompt: str) -> SubagentHandle:
    """Run a copy of this session, in parallel, on something it will not do itself.

    The copy has the same model, the same tools and a private copy of the whole conversation, so
    `prompt` is the difference rather than a briefing.

    The copy starts once this turn's results are recorded, so a later turn is the earliest a wait can
    collect it, and waiting on it in the program that made it never returns it.

    Args:
        prompt: What the copy is to do instead. It has the whole conversation already, so this is the
            difference rather than a briefing.

    Returns:
        The copy's handle, carrying the id a later wait collects it by.

    Raises:
        ApiError: `invalid-argument` for a blank prompt, `limit-exceeded` at the delegation depth
            cap, and `unavailable` when the run has no delegation runtime to run the copy in.
    """
    return _handle(_call(wire.fork, prompt))


__getattr__ = missing(__name__, __all__)
"""What this module answers for a name it does not declare — see `gg._registry.missing`."""

File: packages/gg-sandbox-ruby/src/gg/delegation.rb
# frozen_string_literal: true

module GG
  # Delegate work to child agents, and hand this session's own turn to another agent.
  #
  # Waiting on children blocks while they run, and the run's wall-clock budget keeps ticking
  # throughout.
  #
  # A child is briefed with exactly one of a self-contained `prompt` or a board `issue_id`, both
  # keyword arguments; neither and both are refused by name.
  module Delegation
    extend Surface::Operations

    # The membrane's handle record, as the model-facing one.
    #
    # @param handle [Object] the wire record
    # @return [GG::Delegation::SubagentHandle] the handle a program reads
    # @api private
    def self.handle(handle)
      SubagentHandle.new(
        id: Wire.field(handle, "id"),
        slot: Wire.field(handle, "slot"),
        model_id: Wire.field(handle, "modelId")
      )
    end
    private_class_method :handle

    # Lower a brief onto the membrane's variant.
    #
    # The check is explicit rather than delegated to the bindings because "neither" is the mistake
    # that actually happens — the native JSON schema declares both fields optional and enforces the
    # choice only at dispatch — and the message that comes back has to name both options.
    #
    # @param fn [String] the gg call the failure is reported against
    # @param prompt [String, nil] self-contained instructions
    # @param issue_id [String, nil] a board issue's id
    # @return [Object] the membrane's `subagent-brief` variant
    # @raise [GG::Core::ApiError] `invalid-argument` when neither or both were given
    # @api private
    def self.brief(fn, prompt, issue_id)
      if prompt.nil? == issue_id.nil?
        raise Core::ApiError.new(fn, Core::ApiErrorCode::INVALID_ARGUMENT,
                                 "expected exactly one of `prompt` or `issue_id`")
      end

      prompt.nil? ? Wire.variant("issue", issue_id) : Wire.variant("prompt", prompt)
    end
    private_class_method :brief

    # Delegate scoped work to a child agent and hand back its handle immediately.
    #
    # The child runs in parallel while the program continues, and shares the workspace. `agent`
    # names one of the agent profiles this run declares — the system prompt lists them, and the
    # profile selects the child's model, tools and instructions. The brief is exactly one of
    # `prompt` and `issue_id`.
    #
    # @param agent [String] The agent profile to run the child as, from the ones the system prompt
    #   lists. It selects the child's model, tools and instructions.
    # @param prompt [String, nil] Self-contained instructions for the child. Give this or
    #   `issue_id`, never both and never neither.
    # @param issue_id [String, nil] The board issue to brief the child from. Give this or `prompt`,
    #   never both and never neither.
    # @return [GG::Delegation::SubagentHandle] the child that is now running
    # @raise [GG::Core::ApiError] `:limit_exceeded` at the delegation depth cap, and
    #   `:invalid_argument` when `agent` is not one this run declares.
    def self.spawn_subagent(agent, prompt: nil, issue_id: nil)
      handle(Wire.call("spawn_subagent", "delegation", "spawnSubagent", [
                         Wire.record("agent" => agent,
                                     "task" => brief("spawn_subagent", prompt, issue_id))
                       ]))
    end
    operation :spawn_subagent, "delegation.spawn_subagent", tool: "spawn_subagent"

    # Block until the named children have finished and collect their results in dispatch order.
    #
    # With no arguments it waits for every outstanding child. The run's wall-clock budget keeps
    # running throughout.
    #
    # @param ids [Array<String>] The children to wait for, splatted. Pass none to wait for every one
    #   still outstanding.
    # @return [Array<GG::Delegation::SubagentResult>] what each child finished with, in dispatch
    #   order
    # @raise [GG::Core::ApiError] `:not_found` for an id this session did not spawn.
    def self.wait_for_subagents(*ids)
      wanted = Check.strings("wait_for_subagents", "ids", ids)
      results = Wire.call("wait_for_subagents", "delegation", "waitForSubagents",
                          [wanted.empty? ? `undefined` : wanted])
      results.map do |result|
        SubagentResult.new(
          id: Wire.field(result, "id"),
          status: Wire.symbol(Wire.field(result, "status")),
          summary: Wire.field(result, "summary")
        )
      end
    end
    operation :wait_for_subagents, "delegation.wait_for_subagents", tool: "wait_for_subagents"

    # Deliver a message to a running child agent's inbox, which it reads at its next turn.
    #
    # @param agent_id [String] The child to deliver to, by the id its handle carries.
    # @param message [String] What to put in its inbox. It reads it at its next turn.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:not_found` for an unknown agent id, and `:conflict` when that
    #   child has already returned.
    def self.send_message(agent_id, message)
      Wire.call("send_message", "delegation", "sendMessage", [agent_id, message])
      nil
    end
    operation :send_message, "delegation.send_message", tool: "send_message"

    # Move the process this session is running inside on to another of its states.
    #
    # The state is named the way an agent to spawn is named. It is bound only when a state machine
    # is driving the session and the current state has somewhere to go. It is registered rather
    # than performed: the call validates the target, returns, and the program runs on to its end.
    # The first declaration in a turn is the one that stands.
    #
    # @param state [String] The state to move on to, named the way an agent to spawn is named.
    # @param note [String, nil] The opening message the next state's agent sees. Leave it out to
    #   tell it nothing.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:invalid_argument` for a state this session may not move to, and
    #   `:refused` for a second declaration in one turn.
    def self.transition_state(state, note = nil)
      Wire.call("transition_state", "delegation", "transitionState", [state, Wire.js(note)])
      nil
    end
    operation :transition_state, "delegation.transition_state", tool: "transition_state"

    # Continue this session as a different agent, from the next turn.
    #
    # The named agent takes over with its own model, tools and instructions, keeping every
    # capability the two of them share and the whole conversation. Registered rather than performed:
    # the call validates the target, returns, and the program runs on to its end. A session makes
    # one succession per turn.
    #
    # @param agent [String] The agent to become, from the ones this run declares it may become.
    # @param prompt [String, nil] Its opening message. It already has the whole conversation.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:refused` for a second succession in one turn or one after a
    #   state transition, and `:invalid_argument` for an agent this session may not become.
    def self.exec(agent, prompt = nil)
      Wire.call("exec", "delegation", "exec", [agent, Wire.js(prompt)])
      nil
    end
    operation :exec, "delegation.exec", tool: "exec"

    # Run a copy of this session, in parallel, on something it will not do itself.
    #
    # The copy has the same model, the same tools and a private copy of the whole conversation, so
    # `prompt` is the difference rather than a briefing.
    #
    # Its handle comes back immediately; the copy starts once this turn's tool results are recorded.
    # A later turn is the earliest a wait can collect it, and waiting on it in the program that made
    # it never returns it.
    #
    # @param prompt [String] What the copy is to do instead. It has the whole conversation already.
    # @return [GG::Delegation::SubagentHandle] the copy that starts once this turn is recorded
    # @raise [GG::Core::ApiError] `:limit_exceeded` at the delegation depth cap, and
    #   `:invalid_argument` for a blank prompt.
    def self.fork(prompt)
      handle(Wire.call("fork", "delegation", "fork", [prompt]))
    end
    operation :fork, "delegation.fork", tool: "fork"

    # A child agent that was spawned and is now running in parallel.
    class SubagentHandle
      include Value
      extend Surface::Operations

      # @return [String] The child's id — what `GG::Delegation.wait_for_subagents` and
      #   `GG::Delegation.send_message` take.
      attr_reader :id

      # @return [String] The agent profile it runs as.
      attr_reader :slot

      # @return [String] The model actually bound to that agent.
      attr_reader :model_id

      # @api private
      def initialize(id:, slot:, model_id:)
        @id = id
        @slot = slot
        @model_id = model_id
        freeze
      end

      # Deliver a message to this child's inbox, which it reads at its next turn.
      #
      # @param message [String] What to put in its inbox.
      # @return [nil]
      # @raise [GG::Core::ApiError] `:conflict` when this child has already returned.
      def send_message(message)
        Delegation.send_message(@id, message)
      end
      member_operation :send_message, "delegation.send_message", tool: "send_message"
    end

    # How a child agent's loop ended.
    #
    # Every arm is a Symbol: `GG::Delegation::AgentEnding::COMPLETED` and `:completed` are the same
    # value.
    module AgentEnding
      # It finished normally, ending its own session, and its summary is what it returned.
      COMPLETED = :completed

      # It reached the per-run turn ceiling.
      EXHAUSTED = :exhausted

      # It passed its wall-clock deadline.
      TIMED_OUT = :timed_out

      # A model turn failed.
      MODEL_ERROR = :model_error

      # The run's credential was refused.
      AUTH_ERROR = :auth_error

      # An execution ceiling stopped it — consecutive errors, error rate, or cost.
      LIMIT_EXCEEDED = :limit_exceeded
    end

    # One child agent's collected result.
    class SubagentResult
      include Value

      # @return [String] The child's id.
      attr_reader :id

      # @return [GG::Delegation::AgentEnding, nil] How it finished; `nil` when it produced no return
      #   value at all.
      attr_reader :status

      # @return [String] Its final message.
      attr_reader :summary

      # @api private
      def initialize(id:, status:, summary:)
        @id = id
        @status = status
        @summary = summary
        freeze
      end
    end
  end
end
